ls1intum/Artemis
# `Assessment`: Support long feedback messages

# Contains database migrations

### Checklist
#### General


#### Server
#### Client
#### Changes affecting Programming Exercises

### Motivation and Context
Closes #6296.

### Description
- Adds a new table for the long feedback stored separately from the regular feedback to allow for lazy-loading the long feedback only if actually needed.
- The client only loads the long feedback in case the actual feedback text is un-collapsed. The latest requests are also cached, so that opening and closing the feedback item does not repeatedly fetch the data from the server again.
- For very long feedback a download link to get the full feedback is shown together with the truncated version.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Programming Exercise

Stacked PR on top of #6384. If you can still access the online editor, the changes from #6384 have also been successfully tested.

1. As instructor, create a test case that produces a feedback longer than 5000 characters. E.g.
   ```java
   @Test
   void longFeedback() {
       assertEquals("a".repeat(1000), "1234567890".repeat(500));
   }
   ```
2. Check the result for the template or solution repository. The feedback should contain the full message, i.e. it ends with `…1234567890>` instead of being cut off somewhere in the middle of the sequence.
3. Increase the `500` of the repeat to something like `2000`. Now the feedback is too long to be displayed fully in the UI, instead a download link should be shown.
4. In the network tab of the browser dev console: When you un-collapse the feedback message for the first time, a network request to `long-feedback` should be made. After collapsing, or even closing the feedback detail dialogue and opening it again, the request is not made again.

*Note*: The feedback message for Gradle exercises is cut in half on Jenkins setups due to an unrelated bug currently. → #6434

### Review Progress






#### Code Review
#### Manual Tests

### Test Coverage



| Class/File | Branch | Line |
|------------|-------:|-----:|
| `Feedback.java` | 83% | 97% |
| `LongFeedbackTextResource.java` | 100% | 100% |
| `feedback-text.component.ts` | 100% | 100% |


### Screenshots

![Screenshot 2023-03-26 at 11-52-31 Code Editor](https://user-images.githubusercontent.com/64250573/227768409-7c3e80b5-2694-4306-8bf1-a34f78402f1c.png)

Download link for very long feedback:
![Screenshot 2023-03-26 at 11-52-44 Code Editor](https://user-images.githubusercontent.com/64250573/227768411-fda8b51f-e5f2-4169-b2b0-50101d0b99ce.png)
